File: chapters/03_session_and_dm_scope/v3_scope_policy.py
from __future__ import annotations


def build_session_key(channel: str, user: str, is_dm: bool, policy: str) -> str:
    if policy == "main":
        return "main"

    if policy == "per-peer":
        if is_dm:
            return f"peer:{user}"
        return "main"

    if policy == "per-channel-peer":
        return f"{channel}:{user}"

    raise ValueError(f"unknown policy: {policy}")


def route_session(envelope: dict[str, object], policy: str) -> dict[str, str]:
    channel = str(envelope.get("channel", "unknown"))
    user = str(envelope.get("user", "unknown"))
    is_dm = bool(envelope.get("is_dm", False))

    key = build_session_key(channel, user, is_dm, policy)
    return {
        "policy": policy,
        "session_key": key,
    }


def main() -> None:
    env = {"channel": "telegram", "user": "alice", "is_dm": True}
    print(route_session(env, policy="per-peer"))
    print(route_session(env, policy="per-channel-peer"))


if __name__ == "__main__":
    main()

File: chapters/03_session_and_dm_scope/v3_scope_policy.ts
// TypeScript mirror for Chapter 03: Session and DM Scope (v3)
// Run with: npx ts-node v3_scope_policy.ts

type Policy = "main" | "per-peer" | "per-channel-peer";

interface Envelope {
  channel?: string;
  user?: string;
  is_dm?: boolean;
  [key: string]: unknown;
}

function buildSessionKey(
  channel: string,
  user: string,
  isDm: boolean,
  policy: Policy
): string {
  if (policy === "main") {
    return "main";
  }

  if (policy === "per-peer") {
    if (isDm) {
      return `peer:${user}`;
    }
    return "main";
  }

  if (policy === "per-channel-peer") {
    return `${channel}:${user}`;
  }

  throw new Error(`unknown policy: ${policy}`);
}

function routeSession(
  envelope: Envelope,
  policy: Policy
): { policy: Policy; session_key: string } {
  const channel = envelope.channel ?? "unknown";
  const user = envelope.user ?? "unknown";
  const isDm = envelope.is_dm ?? false;

  const key = buildSessionKey(channel, user, isDm, policy);
  return {
    policy,
    session_key: key,
  };
}

function main(): void {
  const env: Envelope = { channel: "telegram", user: "alice", is_dm: true };
  console.log(routeSession(env, "per-peer"));
  console.log(routeSession(env, "per-channel-peer"));
}

main();
